pageserver/src/tenant/timeline.rs

```diff
@@ -2819,6 +2838,31 @@ impl Timeline {
             },
         )?;
 
+        // Determine N largest holes where N is number of compacted layers
+        let max_holes = deltas_to_compact.len();
+        let min_hole_length = (target_file_size / page_cache::PAGE_SZ as u64) as i128;
+
+        // min-heap (reserve space for one more element added before eviction)
+        let mut heap: BinaryHeap<Hole> = BinaryHeap::with_capacity(max_holes + 1);
+        let mut prev: Option<Key> = None;
+        for (next_key, _next_lsn, _size) in itertools::process_results(
+            deltas_to_compact.iter().map(|l| l.key_iter(ctx)),
+            |iter_iter| iter_iter.kmerge_by(|a, b| a.0 <= b.0),
+        )? {
+            if let Some(prev_key) = prev {
+                if next_key.to_i128() - prev_key.to_i128() >= min_hole_length {
+                    heap.push(Hole(prev_key..next_key));
+                    if heap.len() > max_holes {
+                        heap.pop(); // remove smallest hole
+                    }
+                }
+            }
+            prev = Some(next_key.next());
+        }
+        let mut holes = heap.into_vec();
+        holes.sort_by_key(|hole| hole.0.start);
```

Comment:
`sort_unstable_by_key` would probably work just as well because these are just ranges and we don't need the stable sort.. Probably no difference.

Reply:
Still valid.